File: src/workflows/damage.ts
/**
 * Damage Reporting Workflow
 * @module workflows/damage
 */

import { WizardRenderer } from '../ui/widgets/wizard';
import { getAllLockers, updateLockerDamageStatus } from '../db/lockers';
import { createOptions, createYesNoOptions } from './common';
import type { WizardAction } from '../types/ui';
import type { Locker } from '../types/database';

/**
 * Damage workflow state
 */
export interface DamageWorkflowState {
  step: 'select' | 'action' | 'notes' | 'confirm' | 'complete';
  selectedLocker: Locker | null;
  action: 'report' | 'repair' | null;
  notes: string | null;
}

/**
 * Create initial damage workflow state
 */
export function createDamageWorkflowState(): DamageWorkflowState {
  return {
    step: 'select',
    selectedLocker: null,
    action: null,
    notes: null,
  };
}

/**
 * Damage Reporting Workflow class
 */
export class DamageWorkflow {
  private wizard: WizardRenderer;
  private state: DamageWorkflowState;
  private allLockers: Locker[] = [];

  constructor() {
    this.wizard = new WizardRenderer();
    this.state = createDamageWorkflowState();
    this.allLockers = getAllLockers();
    this.startWorkflow();
  }

  /**
   * Start the workflow
   */
  private startWorkflow(): void {
    this.wizard.addInfo('Schadensmeldung / Reparatur');
    this.showSelectStep();
  }

  /**
   * Show locker selection step
   */
  private showSelectStep(): void {
    if (this.allLockers.length === 0) {
      this.wizard.addInfo('Keine Schließfächer vorhanden.');
      this.state.step = 'complete';
      return;
    }

    const options = createOptions(
      this.allLockers.map((locker) => ({
        label: `${locker.number} (${locker.location}) ${locker.is_damaged ? '⚠️ beschädigt' : '✓ intakt'}`,
        value: locker.id.toString(),
        metadata: { locker },
      }))
    );

    this.wizard.addQuestion('Schließfach auswählen:', options);
    this.state.step = 'select';
  }

  /**
   * Show action selection step
   */
  private showActionStep(): void {
    const locker = this.state.selectedLocker!;

    const options = locker.is_damaged
      ? createOptions([
          { label: 'Als repariert markieren', value: 'repair' },
          { label: 'Abbrechen', value: 'cancel' },
        ])
      : createOptions([
          { label: 'Schaden melden', value: 'report' },
          { label: 'Abbrechen', value: 'cancel' },
        ]);

    this.wizard.addQuestion(
      locker.is_damaged ? 'Schließfach repariert?' : 'Aktion auswählen:',
      options
    );
    this.state.step = 'action';
  }

  /**
   * Show confirmation step
   */
  private showConfirmStep(): void {
    const locker = this.state.selectedLocker!;

    this.wizard.addInfo(`--- ${this.state.action === 'report' ? 'Schadensmeldung' : 'Reparatur'} ---`);
    this.wizard.addInfo(`Schließfach: ${locker.number}`);
    this.wizard.addInfo(`Standort: ${locker.location}`);
    this.wizard.addInfo(`Aktion: ${this.state.action === 'report' ? 'Als beschädigt markieren' : 'Als repariert markieren'}`);

    this.wizard.addQuestion('Bestätigen?', [
      { label: 'Bestätigen', value: 'confirm' },
      { label: 'Abbrechen', value: 'cancel' },
    ]);
    this.state.step = 'confirm';
  }

  /**
   * Execute the damage action
   */
  private executeDamageAction(): void {
    const locker = this.state.selectedLocker!;
    const isDamaged = this.state.action === 'report';

    try {
      updateLockerDamageStatus(locker.id, isDamaged, this.state.notes || undefined);

      if (isDamaged) {
        this.wizard.addInfo('✓ Schaden wurde gemeldet!', 'system');
        this.wizard.addInfo('Schließfach ist jetzt als beschädigt markiert.');
      } else {
        this.wizard.addInfo('✓ Reparatur wurde vermerkt!', 'system');
        this.wizard.addInfo('Schließfach ist wieder verfügbar.');
      }
    } catch (error) {
      this.wizard.addInfo(`✗ Fehler: ${error}`, 'system');
    }

    this.state.step = 'complete';
  }

  /**
   * Handle wizard action
   * @param action Wizard action
   * @returns True if workflow is complete
   */
  handleAction(action: WizardAction): boolean {
    if (action.type === 'cancel') {
      return true;
    }

    if (action.type !== 'next') {
      return false;
    }

    const selected = this.wizard.getSelectedOption();
    if (!selected) return false;

    switch (this.state.step) {
      case 'select':
        const lockerId = parseInt(selected.value, 10);
        this.state.selectedLocker = this.allLockers.find((l) => l.id === lockerId) || null;
        this.showActionStep();
        break;

      case 'action':
        if (selected.value === 'cancel') {
          return true;
        }
        this.state.action = selected.value as 'report' | 'repair';
        this.showConfirmStep();
        break;

      case 'confirm':
        if (selected.value === 'confirm') {
          this.executeDamageAction();
        }
        return true;

      case 'complete':
        return true;
    }

    return false;
  }

  /**
   * Handle key input
   * @param key Key pressed
   * @returns True if workflow is complete
   */
  handleKey(key: string): boolean {
    const action = this.wizard.handleKey(key);
    return this.handleAction(action);
  }

  /**
   * Render the workflow
   * @param width Available width
   * @param height Available height
   * @returns Rendered lines
   */
  render(width: number, height: number): string[] {
    return this.wizard.render(width, height);
  }

  /**
   * Check if workflow is complete
   */
  isComplete(): boolean {
    return this.state.step === 'complete';
  }
}

File: src/workflows/damage.rs
use crate::db::connection::Database;
use crate::{
    db,
    ui::widgets::wizard::{
        MessageContent, MessageSender, WizardAction, WizardMessage, WizardOption, WizardRenderer,
    },
    workflows::common::{WorkflowAction, WorkflowResult},
};
use crossterm::event::KeyCode;

/// Workflow for reporting or repairing locker damage.
pub struct DamageWorkflow {
    renderer: WizardRenderer,
    lockers: Vec<WizardOption>,
    selected_locker: Option<i64>,
    step: u8,
}

impl DamageWorkflow {
    pub fn new(db: &Database) -> Self {
        let lockers = db::lockers::list_lockers(db.connection())
            .unwrap_or_default()
            .into_iter()
            .map(|locker| WizardOption {
                label: format!("{} ({})", locker.number, locker.location),
                value: locker.id.to_string(),
                metadata: Some(if locker.is_damaged { "defekt" } else { "ok" }.to_string()),
            })
            .collect::<Vec<_>>();

        let message = WizardMessage {
            sender: MessageSender::System,
            content: MessageContent::Question {
                text: "Welches Schließfach ist betroffen?".to_string(),
                options: if lockers.is_empty() {
                    vec![WizardOption {
                        label: "Keine Schließfächer verfügbar".to_string(),
                        value: "0".to_string(),
                        metadata: None,
                    }]
                } else {
                    lockers.clone()
                },
            },
            timestamp: None,
        };

        Self {
            renderer: WizardRenderer {
                messages: vec![message],
                current_question_index: 0,
                selected_option_index: 0,
            },
            lockers,
            selected_locker: None,
            step: 0,
        }
    }

    pub fn renderer(&self) -> &WizardRenderer {
        &self.renderer
    }

    pub fn renderer_mut(&mut self) -> &mut WizardRenderer {
        &mut self.renderer
    }

    /// Handles keyboard input for the workflow.
    pub fn handle_key(&mut self, key: KeyCode) -> WorkflowAction {
        match self.renderer.handle_key(key) {
            WizardAction::Confirm(index) => self.handle_confirm(index),
            WizardAction::Cancel => WorkflowAction::Cancelled,
            WizardAction::None => WorkflowAction::None,
        }
    }

    /// Handles selection confirmation for the current step.
    fn handle_confirm(&mut self, index: usize) -> WorkflowAction {
        if self.step == 0 {
            if let Some(option) = self.lockers.get(index) {
                if let Ok(locker_id) = option.value.parse::<i64>() {
                    self.selected_locker = Some(locker_id);
                }
                self.renderer.messages.push(WizardMessage {
                    sender: MessageSender::User,
                    content: MessageContent::Answer {
                        text: option.label.clone(),
                    },
                    timestamp: None,
                });
                self.renderer.messages.push(WizardMessage {
                    sender: MessageSender::System,
                    content: MessageContent::Question {
                        text: "Aktion auswählen".to_string(),
                        options: vec![
                            WizardOption {
                                label: "Defekt melden".to_string(),
                                value: "damage".to_string(),
                                metadata: None,
                            },
                            WizardOption {
                                label: "Repariert".to_string(),
                                value: "repair".to_string(),
                                metadata: None,
                            },
                        ],
                    },
                    timestamp: None,
                });
                self.renderer.current_question_index = self.renderer.messages.len() - 1;
                self.renderer.selected_option_index = 0;
                self.step = 1;
            }
            return WorkflowAction::None;
        }

        if self.step == 1 {
            if let Some(locker_id) = self.selected_locker {
                let damaged = match index {
                    0 => true,
                    _ => false,
                };
                return WorkflowAction::Completed(WorkflowResult::Damage { locker_id, damaged });
            }
        }

        WorkflowAction::None
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::db::{self, connection::Database};

    #[test]
    fn damage_workflow_marks_selection() {
        let db = Database::open_in_memory().expect("db");
        let locker = crate::model::Locker::new(1, "A-01", "Hauptgebäude", "Klein");
        db::lockers::upsert_locker(db.connection(), &locker).expect("locker");

        let mut workflow = DamageWorkflow::new(&db);
        assert!(matches!(
            workflow.handle_key(KeyCode::Enter),
            WorkflowAction::None
        ));
        let result = workflow.handle_key(KeyCode::Enter);
        assert!(matches!(result, WorkflowAction::Completed(_)));
    }
}
